### ISSUE TITLE
Applying a mask with only one of joined cgroup controllers results in being placed in the wrong cgroup

### ISSUE THREAD

**Issue by cdown** [systemd/systemd#10580]
**systemd version the issue has been seen with**
> HEAD

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**

If JoinController=x,y is set and we set `CGROUP_MASK_X` for a unit, the unit is still added to the `x,y` hierarchy with its own service cgroup.

**Unexpected behaviour you saw**

If JoinController=x,y is set and we set `CGROUP_MASK_X` for a unit, the unit gets put in the wrong cgroup in the `x,y` hierarchy (its slice parent).

**Steps to reproduce the problem**

See https://github.com/systemd/systemd/pull/10507#issuecomment-434433621 for where this was found:

1. Use hybrid hierarchy
2. Set `JoinControllers=cpu,cpuacct` (part of the default)
3. Change systemd code so it only applies `CGROUP_MASK_CPU` for `CPUShares` in src/core/cgroup.c in cgroup_context_has_cpu_weight/cgroup_context_has_cpu_shares check at the top of cgroup_context_get_mask
4. `systemd-run -p CPUShares=100 sleep infinity`
5. Look at cgroups for `sleep infinity` process (see cpu,cpuacct hierarchy where we've been placed in parent slice):

```
12:pids:/system.slice/run-r2969b148d2cd4305989b0573144eb674.service
11:blkio:/
10:cpu,cpuacct:/system.slice
9:devices:/system.slice/run-r2969b148d2cd4305989b0573144eb674.service
8:net_cls,net_prio:/
7:freezer:/
6:hugetlb:/
5:cpuset:/
4:rdma:/
3:memory:/system.slice/run-r2969b148d2cd4305989b0573144eb674.service
2:perf_event:/
1:name=systemd:/system.slice/run-r2969b148d2cd4305989b0573144eb674.service
0::/system.slice/run-r2969b148d2cd4305989b0573144eb674.service
```

My guess is that this deficiency went unnoticed because we only usually apply cpuacct and cpu together in old code, and net_cls/net_prio was not so commonly exercised independently either.

**Comment by poettering** [comment]
arks... hmm, so this is not surprising. After all if the unit is in one cgroup in one jointly mounted contorller and in another cgroup in another controller mounted in the same hierarchy, then this just means that we'll first add the process to a cgroup 1 and then move it to cgroup 2 right away... so this behaviour is really not surprising...

i am kinda tempted to remove the ability to configure controllers mounted together freely and instead simply hardcode that net_cls/net_prio are always mounted together and cpuacct/cpu, so that it is easy for us to determine the "deepest" cgroup per jointly mounted controller.

What a mess. and kinda amazing that nobody noticed how fucked this was earlier.

**Comment by cdown** [comment]
Happy to take a look at fixing this after #10507 is sorted. I think removing free choice with JoinControllers is probably a reasonable idea. It also allows us to clean up the masks which are no longer used in unified.

**Comment by poettering** [comment]
Fixed by #10785 (well, that PR removes the whole concept, hence the problem set goes away entirely)

**Comment by poettering** [comment]
#10785 was merged now. Closing. (Forgot to add the tag to that commit msg...)
